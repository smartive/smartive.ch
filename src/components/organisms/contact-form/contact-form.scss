@import '../../../scss/colors';
@import '../../../scss/responsive';

.contact-form {
  .button {
    width: 100%;

    @media #{$small-up} {
      width: 250px;

      &[name='goal'] {
        width: 400px;
      }
    }
  }

  form {
    overflow: hidden;
    word-break: break-word;
  }

  .inline-input {
    color: $smartive-dark-blue;
    border: none;
    border-bottom: 2px dotted $light-gray;
    background: none;
    font-size: 1em;
    padding: 0;
    outline: none;
  }

  .inline-input-placeholder {
    border: none;
    font-size: 1em;
    background: none;
    padding: 0;
    font-weight: inherit;
    font-family: inherit;
    cursor: text;
    outline: none;
    color: $smartive-blue;
    border-bottom: 2px dotted $smartive-blue;
  }

  &__message {
    height: 2rem;
    padding-top: 0.5rem;
    font-size: 0.8rem;

    &--error {
      color: $red;
    }
    &--success {
      color: $green;
    }
  }
}
