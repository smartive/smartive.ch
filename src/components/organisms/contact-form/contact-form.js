import React, { useState } from 'react';
import { Button } from '../../atoms';
import { InlineInput } from './inline-input';

import './contact-form.scss';

const DEFAULT_ERROR_MESSAGE = 'Es ist ein Fehler aufgetreten. Bitte versuchen Sie es noch einmal.';
const THANKS_MESSAGE = 'Danke für Ihre Nachricht. Wir werden uns bei Ihnen melden.';
const FORM_URL = 'https://formspree.io/xnqgjbjz';

export const ContactForm = () => {
  const [subject, setSubject] = useState('');
  const [additionalInfo, setAdditionalInfo] = useState('');
  const [name, setName] = useState('');
  const [contact, setContact] = useState('');

  const [loading, setLoading] = useState(false);
  const [message, setMessage] = useState(null);

  const resetForm = () => {
    setSubject('');
    setAdditionalInfo('');
    setName('');
    setContact('');
  };

  const submit = async event => {
    event.preventDefault();

    if (loading) {
      return;
    }

    if (!contact.trim() || !name.trim() || !subject.trim()) {
      setMessage({ type: 'error', text: 'Bitte mindestens Name, E-Mail Adresse und Anliegen angeben.' });
      return;
    }

    setLoading(true);

    try {
      await fetch(FORM_URL, {
        method: 'POST',
        body: JSON.stringify({ name, contact, subject, additionalInfo }),
      });
      setMessage({ type: 'success', text: THANKS_MESSAGE });
      resetForm();
    } catch (error) {
      setMessage({ type: 'error', text: DEFAULT_ERROR_MESSAGE });
    } finally {
      setLoading(false);
    }
  };

  return (
    <div className="contact-form">
      <p>Verraten Sie uns, worüber Sie gern mit uns sprechen möchten.</p>
      <form onSubmit={submit}>
        <p>
          Ich heisse <InlineInput setValue={setName} placeholderText="mein Name" />. Gerne würde ich mich mit Euch über{' '}
          <InlineInput setValue={setSubject} placeholderText="mein Thema" /> austauschen. Dabei geht es mir vor allem um{' '}
          <InlineInput setValue={setAdditionalInfo} placeholderText="zusätzliche Infos" />. Am besten erreicht ihr mich unter{' '}
          <InlineInput setValue={setContact} placeholderText="meine E-Mail / Tel." />.
        </p>

        <Button text={loading ? 'Sende...' : 'Senden'} disabled={loading} onClick={submit} />
        <div className={`contact-form__message ${message ? `contact-form__message--${message.type}` : ''}`}>
          {message && message.text}
        </div>
      </form>
    </div>
  );
};

export default ContactForm;
