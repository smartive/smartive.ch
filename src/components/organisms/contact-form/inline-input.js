import PropTypes from 'prop-types';
import React, { useRef } from 'react';

import './contact-form.scss';

export const InlineInput = ({ setValue, placeholderText }) => {
  const inputRef = useRef(null);
  const placeholderRef = useRef(null);

  return (
    <>
      <span
        className="inline-input"
        role="textbox"
        ref={inputRef}
        aria-label={placeholderText}
        contentEditable
        onInput={({ target }) => {
          setValue(target.innerHTML || '');

          if (placeholderRef.current) {
            placeholderRef.current.style.display = target.innerHTML ? 'none' : 'inline';
          }
        }}
      />
      <button
        ref={placeholderRef}
        className="inline-input-placeholder"
        type="button"
        onClick={() => {
          if (placeholderRef.current && inputRef.current) {
            inputRef.current.focus();
          }
        }}
      >
        {placeholderText}
      </button>
    </>
  );
};

InlineInput.propTypes = {
  setValue: PropTypes.func.isRequired,
  placeholderText: PropTypes.string.isRequired,
};

export default InlineInput;
