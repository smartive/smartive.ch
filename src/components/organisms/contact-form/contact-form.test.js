import React from 'react';
import renderer from 'react-test-renderer';

import { ContactForm } from './contact-form';

describe('ContactForm', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<ContactForm />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
