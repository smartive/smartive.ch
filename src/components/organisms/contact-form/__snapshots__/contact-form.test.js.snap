// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm renders correctly 1`] = `
<div
  className="contact-form"
>
  <p>
    Verraten Sie uns, worüber Sie gern mit uns sprechen möchten.
  </p>
  <form
    onSubmit={[Function]}
  >
    <p>
      Ich heisse 
      <span
        aria-label="mein Name"
        className="inline-input"
        contentEditable={true}
        onInput={[Function]}
        role="textbox"
      />
      <button
        className="inline-input-placeholder"
        onClick={[Function]}
        type="button"
      >
        mein Name
      </button>
      . Gerne würde ich mich mit Euch über
       
      <span
        aria-label="mein Thema"
        className="inline-input"
        contentEditable={true}
        onInput={[Function]}
        role="textbox"
      />
      <button
        className="inline-input-placeholder"
        onClick={[Function]}
        type="button"
      >
        mein Thema
      </button>
       austauschen. Dabei geht es mir vor allem um
       
      <span
        aria-label="zusätzliche Infos"
        className="inline-input"
        contentEditable={true}
        onInput={[Function]}
        role="textbox"
      />
      <button
        className="inline-input-placeholder"
        onClick={[Function]}
        type="button"
      >
        zusätzliche Infos
      </button>
      . Am besten erreicht ihr mich unter
       
      <span
        aria-label="meine E-Mail / Tel."
        className="inline-input"
        contentEditable={true}
        onInput={[Function]}
        role="textbox"
      />
      <button
        className="inline-input-placeholder"
        onClick={[Function]}
        type="button"
      >
        meine E-Mail / Tel.
      </button>
      .
    </p>
    <button
      className="button"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Senden
    </button>
    <div
      className="contact-form__message "
    />
  </form>
</div>
`;
